### PR TITLE
Complete gate zoo and fix gate docs

### PR DESCRIPTION
**Pull request by dabacon**
Completes work for #852

### CODE REVIEW

File: cirq-core/cirq/ops/boolean_hamiltonian.py [hunk 1]
```diff
@@ -36,7 +36,20 @@
...
 class BooleanHamiltonianGate(raw_types.Gate):
-    """A gate that represents a Hamiltonian from a set of Boolean functions."""
+    r"""A gate that represents evolution due to a Hamiltonian from a set of Boolean functions.
+
+    This gate constructs a diagonal gate in the computational basis that encodes in its
+    phases classical functions.
+
+    The gate is specified by a list of parameters, $[x_0, x_1, \dots, x_{n-1}]$, a
+    list of boolean expressions that are functions of these parameters,
+    $[f_0(x_0,\dots,x_{n-1}), f_1(x_0,\dots,x_{n-1}), f_{p-1}(x_0,\dots,x_{n-1})]$
```
Comment (viathor) [inline cirq-core/cirq/ops/boolean_hamiltonian.py:46]: nit: Missing dots between f_1 and f_{p-1}:

```
$[f_0(x_0,\dots,x_{n-1}), f_1(x_0,\dots,x_{n-1}), \dots, f_{p-1}(x_0,\dots,x_{n-1})]$
```

(If it gets too long, you can remove f_1.)

File: cirq-core/cirq/ops/common_gates.py [hunk 2]
```diff
@@ -906,18 +906,18 @@ def __repr__(self) -> str:
...
-         [0, 1, 0, 0],
-         [0, 0, 1, 0],
-         [0, 0, 0, g]]
-
-    where:
-
-        g = exp(i·π·t).
+    $$
+    \begin{bmatrix}
+        1 & 0 & 0 & 0 \\
```
Comment (viathor) [inline cirq-core/cirq/ops/common_gates.py:915]: This nit applies throughout: A lot of the matrices have few non-zero elements and they are easier to read if the zeros are not explicitly shown:

```
$$
\begin{bmatrix}
    1 & & & \\
    & 1 & & \\
    & & 1 & \\
    & & & e^{i \pi t} \\
\end{bmatrix}
$$

File: cirq-core/cirq/ops/common_gates.py [hunk 3]
```diff
@@ -1322,34 +1332,40 @@ def cphase(rads: value.TParamVal) -> CZPowGate:
 CZ = CZPowGate()
 document(
     CZ,
-    """The controlled Z gate.
+    r"""The controlled Z gate.
 
-    The `exponent=1` instance of `cirq.CZPowGate`.
+    This the `exponent=1` instance of `cirq.CZPowGate`.
```
Comment (viathor) [inline cirq-core/cirq/ops/common_gates.py:1337]: nit: Did you mean to say "This is the..."?

File: cirq-core/cirq/ops/diagonal_gate.py [hunk 4]
```diff
@@ -64,12 +64,16 @@ def _gen_gray_code(n: int) -> Iterator[Tuple[int, int]]:
 
 @value.value_equality()
 class DiagonalGate(raw_types.Gate):
-    """A gate given by a diagonal (2^n)\\times(2^n) matrix."""
+    r"""An n qubit gate which acts as phases on computational basis states.
+
+    This gate's off-diagonal elements are zero and its on diagonal elements are
```
Comment (viathor) [inline cirq-core/cirq/ops/diagonal_gate.py:69]: nit: s/on diagonal/on-diagonal/ or s/on diagonal/diagonal/

(Applies below, too.)

File: cirq-core/cirq/ops/fourier_transform.py [hunk 5]
```diff
@@ -86,7 +98,19 @@ def _circuit_diagram_info_(
 
 @value.value_equality
 class PhaseGradientGate(raw_types.Gate):
-    """Phases each state |k⟩ out of n by e^(2*pi*i*k/n*exponent)."""
+    r"""Phases each all computational basis states proportional to the integer value of the state.
```
Comment (viathor) [inline cirq-core/cirq/ops/fourier_transform.py:101]: nit: s/each all/all/

File: cirq-core/cirq/ops/fourier_transform.py [hunk 6]
```diff
@@ -24,7 +24,19 @@
 
 @value.value_equality
 class QuantumFourierTransformGate(raw_types.Gate):
-    """Switches from the computational basis to the frequency basis."""
+    r"""Switches from the computational basis to the frequency basis.
+
+    This gate has the unitary
+
+    $$
+    \frac{1}{2^{n/2}}\sum_{x,y=0}^{2^n-1} \omega ^{xy} |x\rangle\langle y|
```
Comment (viathor) [inline cirq-core/cirq/ops/fourier_transform.py:32]: nit: unnecessary space between \omega and ^

File: cirq-core/cirq/ops/fourier_transform.py [hunk 7]
```diff
@@ -86,7 +98,19 @@ def _circuit_diagram_info_(
 
 @value.value_equality
 class PhaseGradientGate(raw_types.Gate):
-    """Phases each state |k⟩ out of n by e^(2*pi*i*k/n*exponent)."""
+    r"""Phases each all computational basis states proportional to the integer value of the state.
+
+    The gate `cirq.PhaseGradientGate(n, t)` has the unitary
+    $$
+    \sum_{x=0}^{2^n-1} \omega^t |x\rangle \langle x|
```
Comment (viathor) [inline cirq-core/cirq/ops/fourier_transform.py:105]: s/^t/^x/

File: cirq-core/cirq/ops/fsim_gate.py [hunk 8]
```diff
@@ -202,20 +214,24 @@ def _json_dict_(self) -> Dict[str, Any]:
...
     The unitary matrix of PhasedFSimGate(θ, ζ, χ, γ, φ) is:
 
-        [[1,                       0,                       0,            0],
-         [0,    exp(-iγ - iζ) cos(θ), -i exp(-iγ + iχ) sin(θ),            0],
-         [0, -i exp(-iγ - iχ) sin(θ),    exp(-iγ + iζ) cos(θ),            0],
-         [0,                       0,                       0, exp(-2iγ-iφ)]].
+    $$
+    \begin{bmatrix}
+        1 & 0 & 0 & 0 \\
+        0 & e^{-i \gamma-i \zeta} & -i e^{-i \gamma + i\chi} & 0 \\
```
Comment (viathor) [inline cirq-core/cirq/ops/fsim_gate.py:224]: nit: add spaces around the second minus in the first exponential for consistency with the others

File: cirq-core/cirq/ops/three_qubit_gates.py [hunk 9]
```diff
@@ -49,9 +49,22 @@
...
+    $$
+    \begin{bmatrix}
+        1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\
+        0 & 1 & 0 & 0 & 0 & 0 & 0 & 0 \\
+        0 & 0 & 1 & 0 & 0 & 0 & 0 & 0 \\
+        0 & 0 & 0 & 1 & 0 & 0 & 0 & 0 \\
+        0 & 0 & 0 & 0 & 1 & 0 & 0 & 0 \\
+        0 & 0 & 0 & 0 & 0 & 1 & 0 & 0 \\
+        0 & 0 & 0 & 0 & 0 & 0 & 1 & 0 \\
+        0 & 0 & 0 & 0 & 0 & 0 & 0 & e^{i \pi t}
```
Comment (viathor) [inline cirq-core/cirq/ops/three_qubit_gates.py:65]: (This is a nice example showing the advantage of implicit zeros I advocated for in another comment. It takes some eye-strain to recognize that this is diagonal.)

File: cirq-core/cirq/ops/three_qubit_gates.py [hunk 10]
```diff
@@ -202,12 +215,16 @@ def controlled(
 
 @value.value_equality()
 class ThreeQubitDiagonalGate(raw_types.Gate):
-    """A gate given by a diagonal 8x8 matrix."""
+    r"""A three qubit gate given whose unitary is given by a diagonal $8 \times 8$ matrix.
```
Comment (viathor) [inline cirq-core/cirq/ops/three_qubit_gates.py:218]: s/given whose/whose/

File: cirq-core/cirq/ops/two_qubit_diagonal_gate.py [hunk 11]
```diff
@@ -31,12 +31,28 @@
 
 @value.value_equality()
 class TwoQubitDiagonalGate(raw_types.Gate):
-    """A gate given by a diagonal 4\\times 4 matrix."""
+    r"""A two qubit gate whose unitary is a diagonal $4 \times 4$ matrix.
+
+    This gate's off-diagonal elements are zero and its on diagonal
```
Comment (viathor) [inline cirq-core/cirq/ops/two_qubit_diagonal_gate.py:36]: I think this should say "on-diagonal" or just "diagonal".

(Applies below on l.55, too.)

**Review by viathor — APPROVED** [review]
LGTM with many nits

File: cirq-core/cirq/ops/common_gates.py (hunk 2 above)
Comment (dabacon) [reply cirq-core/cirq/ops/common_gates.py:915]: For this one I'm going to open an issue.  I struggled with this myself.  For three qubit gates the sparse representation is a definite yes.  But I worry a bit that not showing the 0s will confuse users

File: cirq-core/cirq/ops/three_qubit_gates.py (hunk 9 above)
Comment (dabacon) [reply cirq-core/cirq/ops/three_qubit_gates.py]: done here for 3 qubit gates, but left 2 qubit gates in #5359

**Pull request merged**